### PR TITLE
LB refactoring

### PR DESCRIPTION
**Pull request by alena1108**
Cattle changes for lb refactor project: convert all lb legacy rules to the metadata

### CODE REVIEW

**Comment by alena1108** [comment]
https://github.com/rancher/lb-controller/pull/15

**Comment by alena1108** [comment]
https://github.com/rancher/rancher/issues/2179

File: resources/content/cattle-global.properties [hunk 1]
```diff
@@ -6,6 +6,7 @@ api.ui.index=//releases.rancher.com/ui/latest
 api.ui.js.url=https://releases.rancher.com/api-ui/1.0.4/ui.min.js
 api.ui.css.url=https://releases.rancher.com/api-ui/1.0.4/ui.min.css
 agent.instance.image=rancher/agent-instance:v0.8.3
+lb.instance.image=rancher/lb-service-haproxy:v0.3.0
```
Comment (alena1108) [inline resources/content/cattle-global.properties]: This is the default image if nothing is passed to the API.

**Comment by alena1108** [comment]
Only one remaining question - how are we doing to distribute images for lb services. Don't want to hold it in the global-properties as it is going to tight rancher to lb-controller releases. Would have to find a way to distribute it similar way we do for our system stacks - via catalog. Only in this case we can't expect lb to run in its own dedicated stack, as other items deployed from catalog do.

**Comment by alena1108** [comment]
@vincent99 update on per discussion with @ibuildthecloud:

1) we no longer have a special type for the balancersService; regular service type should be used instead. As identifier for LB/CLI/rancher-compose, label "io.rancher.balancer.service"=true will be used.
2) All lb data should be passed on service create/update a part of service.metadata field, under "lb" key.
3) For UI/CLI/rancher-compose, I've created a helper API type - "balancerServiceMetadata" where all lb fields and their types are defined.

File: code/iaas/config-item/server/src/main/java/io/cattle/platform/configitem/context/dao/LoadBalancerInfoDao.java [hunk 2]
```diff
@@ -19,4 +20,8 @@
             List<? extends LoadBalancerListenerInfo> listeners);
 
     List<LoadBalancerTargetInput> getLoadBalancerTargets(Service lbService);
+
+    Map<String, Object> processLBMetadata(Service lbService, LoadBalancerInfoDao lbInfoDao, Map<String, Object> meta);
```
Comment (alena1108) [inline code/iaas/config-item/server/src/main/java/io/cattle/platform/configitem/context/dao/LoadBalancerInfoDao.java]: Two methods added are not used today, but in the future can be used for converting LB service data to "lb" metadata used by the new microservice

File: resources/content/schema/base/balancerTargetMetadata.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
```
Comment (alena1108) [inline resources/content/schema/base/balancerTargetMetadata.json]: this type is for the lb service metadata defined on the target service (when service joins the LB based on selector)

**Comment by alena1108** [comment]
Corresponding catalog PR: https://github.com/rancher/rancher-catalog/pull/208/files

**Comment by vincent99** [comment]
@alena1108 @ibuildthecloud Boo?  Service with a magic label and metadata that happens to match a type isn't discoverable from the API.

File: code/iaas/agent-instance/src/main/java/io/cattle/platform/agent/instance/factory/impl/AgentInstanceFactoryImpl.java [hunk 4]
```diff
@@ -155,6 +155,27 @@ public Agent createAgent(Instance instance) {
         return null;
     }
 
+    @SuppressWarnings("unchecked")
+    protected boolean isSystemService(Service service) {
+        Stack stack = objectManager.loadResource(Stack.class, service.getStackId());
+        if (DataAccessor.fieldBool(stack, "isSystem")) {
```
Comment (ibuildthecloud) [inline code/iaas/agent-instance/src/main/java/io/cattle/platform/agent/instance/factory/impl/AgentInstanceFactoryImpl.java]: Can you use ServiceConstants.isService()

Comment (ibuildthecloud) [inline code/iaas/agent-instance/src/main/java/io/cattle/platform/agent/instance/factory/impl/AgentInstanceFactoryImpl.java]: sorry isSystem()

**Pull request merged**